dbt-materialize: update setup and config guide (#2326)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-configs/materialize-configs.md

## Performance optimizations

### Clusters

<Changelog>

- **v1.2.0:** Enable the configuration of [clusters](https://github.com/MaterializeInc/materialize/blob/main/misc/dbt-materialize/CHANGELOG.md#120---2022-08-31).

</Changelog>

> The [cluster](https://materialize.com/docs/overview/key-concepts/#clusters) that dbt uses to maintain materialized views or indexes can be configured in your [profile](/reference/profiles.yml) for Materialize connections. To override the cluster that is used for specific models (or groups of models), use the `cluster` model configuration.

Reviewer: ```suggestion
The default [cluster](https://materialize.com/docs/overview/key-concepts/#clusters) that is used to maintain materialized views or indexes can be configured in your [profile](/reference/profiles.yml) using the `cluster` connection parameter. To override the cluster that is used for specific models (or groups of models), use the `cluster` configuration parameter.
```

<File name='my_view_cluster.sql'>

```sql
{{ config(materialized='materializedview', cluster='not_default') }}

select ...
```

</File>

<File name='dbt_project.yml'>

```yaml
models:
  project_name:
    +materialized: materializedview
    +cluster: not_default
```

</File>

> **Recommendation** Create a cluster to use for anything built by dbt. This will leave Materialize's default cluster free for responding quickly to `SHOW MATERIALIZED VIEWS` and other diagnostic queries.

Reviewer: ```suggestion
```

Reviewer: > This will leave Materialize's default cluster free for responding quickly to `SHOW MATERIALIZED VIEWS` and other diagnostic queries.

Diagnostic queries should run in the `mz_introspection` cluster, not `default` (also because this cluster can be dropped at any time). We can add recommendations about workflows once we have more palpable user data on what the actual best practices are!

### Incremental models: Materialized Views

Materialize, at its core, is a real-time database that delivers incremental view updates without ever compromising on latency or correctness. Use [materialized views](https://materialize.com/docs/overview/key-concepts/#materialized-views) to compute and incrementally update the results of your query.

### Indexes

<Changelog>

- **v1.2.0:** Enable additional configuration for [indexes](https://github.com/MaterializeInc/materialize/blob/main/misc/dbt-materialize/CHANGELOG.md#120---2022-08-31).

</Changelog>

> Indexes assemble and maintain a query’s results in memory within a cluster, which provides future queries the data they need in a format they can immediately use.

Reviewer: ```suggestion
Like in any standard relational database, you can use [indexes](https://materialize.com/docs/overview/key-concepts/#indexes) to optimize query performance in Materialize. Improvements can be significant, reducing response times down to single-digit milliseconds.
```

Reviewer: It helps to make the description less Materialize-y and more relatable to dbt users (we also need to make it more relatable in our docs; the suggestion is from a recent attempt to do that in the `Optimization` page).

Materialized views (`materializedview`), views (`view`) and sources (`source`) may have a list of `indexes` defined. Each [Materialize index](https://materialize.com/docs/sql/create-index/) can have the following components:

- `columns` (list, required): one or more columns on which the index is defined. To create an index that uses _all_ columns, use the `default` component instead.
- `name` (string, optional): the name for the index. If unspecified, Materialize will use the materialization name and column names provided.
- `cluster` (string, optional): the cluster to use to create the index. If unspecified, indexes will be created in the cluster used to create the materialization.
- `default` (bool, optional): Default: `False`. If set to `True`, creates a default index that uses all columns.

<File name='my_view_index.sql'>

```sql

> {{ config(materialized='view',

Reviewer: ```suggestion
{{ config(materialized='view',
          indexes=[{'columns': ['col_a'], 'cluster': 'cluster_a'}]) }}
```

Reviewer: It's useful to show the `cluster` configuration here.

## Performance optimizations

> ### Clusters
> 
> <Changelog>
> 
> - **v1.2.0:** Enable the configuration of [clusters](https://github.com/MaterializeInc/materialize/blob/main/misc/dbt-materialize/CHANGELOG.md#120---2022-08-31).
> 
> </Changelog>
> 
> The default [cluster](https://materialize.com/docs/overview/key-concepts/#clusters) that is used to maintain materialized views or indexes can be configured in your [profile](/reference/profiles.yml) using the `cluster` connection parameter. To override the cluster that is used for specific models (or groups of models), use the `cluster` configuration parameter.
> 
> <File name='my_view_cluster.sql'>
> 
> ```sql
> {{ config(materialized='materializedview', cluster='not_default') }}
> 
> select ...
> ```
> 
> </File>
> 
> <File name='dbt_project.yml'>
> 
> ```yaml
> models:
>   project_name:
>     +materialized: materializedview
>     +cluster: not_default
> ```
> 
> </File>
> 
> 

Reviewer 2: we have a nifty module specifically for this use case! if you're interested feel free to use it, otherwise we can keep it as it. See [versioning blocks of content](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-blocks-of-content) i think it'd look something like this (but GitHub sucks at rendering suggestions that themselves have triple backticks in them. but docusaurus will render it ok
```
<VersionBlock firstVersion="0.21" lastVersion="1.0">

	your stuff

</VersionBlock>
```

```suggestion
<VersionBlock firstVersion="1.2">
### Clusters

<Changelog>

- **v1.2.0:** Enable the configuration of [clusters](https://github.com/MaterializeInc/materialize/blob/main/misc/dbt-materialize/CHANGELOG.md#120---2022-08-31).

</Changelog>

The default [cluster](https://materialize.com/docs/overview/key-concepts/#clusters) that is used to maintain materialized views or indexes can be configured in your [profile](/reference/profiles.yml) using the `cluster` connection parameter. To override the cluster that is used for specific models (or groups of models), use the `cluster` configuration parameter.

<File name='my_view_cluster.sql'>

```sql
{{ config(materialized='materializedview', cluster='not_default') }}

select ...
```

</File>

<File name='dbt_project.yml'>

```yaml
models:
  project_name:
    +materialized: materializedview
    +cluster: not_default
```

</File>

</VersionBlock>

```
